packages/state-transition/src/cache/epochCache.ts

```diff
@@ -736,9 +768,66 @@ export class EpochCache {
     return isAggregatorFromCommitteeLength(committee.length, slotSignature);
   }
 
+  /**
+   * Return finalized pubkey given the validator index.
+   * Only finalized pubkey as we do not store unfinalized pubkey because no where in the spec has a
+   * need to make such enquiry
+   *
+   */
+  getPubkey(index: ValidatorIndex): PublicKey {
+    return this.index2pubkey[index];
+  }
+
+  getValidatorIndex(pubkey: Uint8Array | PubkeyHex): ValidatorIndex | undefined {
+    if (this.isAfterEIP6110()) {
+      return this.pubkey2index.get(pubkey) ?? this.unfinalizedPubkey2index.get(toMemoryEfficientHexStr(pubkey));
+    } else {
+      return this.pubkey2index.get(pubkey);
+    }
+  }
+
+  /**
+   *
+   * Add unfinalized pubkeys
+   *
+   */
   addPubkey(index: ValidatorIndex, pubkey: Uint8Array): void {
+    if (this.isAfterEIP6110()) {
+      this.unfinalizedPubkey2index = this.unfinalizedPubkey2index.set(toMemoryEfficientHexStr(pubkey), index);
+    } else {
+      this.pubkey2index.set(pubkey, index);
+      this.index2pubkey[index] = bls.PublicKey.fromBytes(pubkey, CoordType.jacobian);
+    }
+  }
+
+  /**
+   *
+   * Given a validator whose activationEligibilityEpoch has just been set, we move its pubkey from unfinalized cache to finalized cache
+   *
+   */
+  addFinalizedPubkey(pubkey: Uint8Array, index: ValidatorIndex): void {
+    if (!this.isAfterEIP6110()) {
+      throw new Error("addFInalizedPubkey is not available pre EIP-6110");
+    }
+
+    const existingIndex = this.pubkey2index.get(pubkey);
+
+    if (existingIndex != undefined) {
+      if (existingIndex === index) {
+        // Repeated insert. Should not happen except during the first few epochs of 6110 activation
+        // Unfinalized validator added to finalizedPubkey2index pre-6110 by calling addPubkey()
+        // when it become finalized in post-6110, addFinalizedPubkey() is called to cause repeated insert
+        return;
+      } else {
+        // attempt to insert the same pubkey with different index, should never happen.
+        throw Error("inserted existing pubkey into finalizedPubkey2index cache with a different index");
+      }
+    }
+
     this.pubkey2index.set(pubkey, index);
-    this.index2pubkey[index] = bls.PublicKey.fromBytes(pubkey, CoordType.jacobian); // Optimize for aggregation
+    this.index2pubkey[index] = bls.PublicKey.fromBytes(pubkey, CoordType.jacobian);
```

Comment:
we may not need to convert from `Uint8Array` to `PublicKey` again, the consumer should have `PublicKey` object already